src/tango/ip/fd_netlink.c

```diff
@@ -0,0 +1,760 @@
+             FD_NL_ARP_FLAGS_IP_ADDR
+             FD_NL_ARP_FLAGS_MAC_ADDR
+             FD_NL_ARP_FLAGS_IFINDEX
+
+         the FD_NL_ARP_FLAGS_UNSUPPORTED flags must not be present */
+      if( ( entry->flags & FD_NL_ARP_FLAGS_UNSUPPORTED ) == 0 ) {
+        uint rqd = FD_NL_ARP_FLAGS_IP_ADDR   |
+          FD_NL_ARP_FLAGS_MAC_ADDR  |
+          FD_NL_ARP_FLAGS_IFINDEX;
+        if( ( entry->flags & rqd ) == rqd ) {
+          entry->flags |= FD_NL_ARP_FLAGS_USED;
+          arp_entry_idx++;
+        }
+      }
+    }
+
+    h = NLMSG_NEXT( h, msglen );
+  }
+
+  /* clear remaining entries */
+  for( long j = arp_entry_idx; j < (long)arp_table_cap; ++j ) {
+    fd_memset( arp_table + j, 0, sizeof( arp_table[0] ) );
+  }
+
+  return arp_entry_idx;
+}
+
+
+fd_nl_arp_entry_t *
+fd_nl_arp_query( fd_nl_arp_entry_t * arp_table,
+                 ulong               arp_table_sz,
+                 uint                ip_addr ) {
+  for( long j = 0L; j < (long)arp_table_sz; ++j ) {
+    fd_nl_arp_entry_t * entry = arp_table + j;
+    if( ( entry->flags & FD_NL_ARP_FLAGS_USED ) == 0 ) break;
+
+    if( entry->dst_ip_addr == ip_addr ) {
+      return entry;
+    }
+  }
+
+  return NULL;
+}
+
+
+int
+fd_nl_update_arp_table( fd_nl_t * nl,
+                        uint      ip_addr,
+                        uint      ifindex ) {
+  int fd = nl->fd;
+  if( fd < 0 ) return -1;
+
+  uint net_ip_addr = htonl( ip_addr );
+
+  /* format the request */
+
+  /* Request struct */
+  struct {
+    struct nlmsghdr nlh;      /* Netlink header */
+    struct ndmsg    ndm;      /* Payload - neighbor message */
+    uchar           buf[128]; /* TODO choose approproate length */
+  } nl_request;
+
+  fd_memset( &nl_request, 0, sizeof( nl_request ) );
+
+  uint seq = nl->seq++;
+
+  nl_request.nlh.nlmsg_type  = RTM_NEWNEIGH;  /* We wish to get neighbors */
+  nl_request.nlh.nlmsg_flags = NLM_F_REQUEST | NLM_F_CREATE | NLM_F_EXCL;
+  nl_request.nlh.nlmsg_len   = NLMSG_LENGTH(sizeof(struct rtmsg));
+  nl_request.nlh.nlmsg_pid   = 0;
+  nl_request.nlh.nlmsg_seq   = seq;
+
+  /* request IPv4 entries */
+  nl_request.ndm.ndm_family  = AF_INET;
+  nl_request.ndm.ndm_ifindex = (int)ifindex;
+  nl_request.ndm.ndm_state   = NUD_INCOMPLETE;
+  nl_request.ndm.ndm_flags   = 0;
+  nl_request.ndm.ndm_type    = RTN_UNICAST;
+
+  /* not necessarily defined! */
+#ifndef NLMSG_TAIL
+#define NLMSG_TAIL(nmsg) ((struct rtattr *)(((uchar *)(nmsg)) + NLMSG_ALIGN((nmsg)->nlmsg_len)))
+#endif
+
+  /* write required attributes */
+  struct rtattr * rqs_rat = NLMSG_TAIL(&nl_request.nlh);
+  // = { .rta_type = NDA_DST, .rta_data_sz = 4 + sizeof( struct rtattr ) };
+
+  rqs_rat->rta_type = NDA_DST;
+  rqs_rat->rta_len =  RTA_LENGTH(4);
+
+  memcpy( RTA_DATA(rqs_rat), &net_ip_addr, 4 );
+  nl_request.nlh.nlmsg_len += sizeof(struct rtattr) + 4;
+
+  ulong nl_request_sz = (ulong)nl_request.nlh.nlmsg_len;
+
+  long sent = send( fd, &nl_request, nl_request_sz, 0 );
```

Comment:
As above, want to check that sent is the expected size or send() in a loop til the buffer is drained